### PR TITLE
Support tuples for type directive tokens

### PR DESCRIPTION
**Pull request by ajaybhargavb**
#1592 

@NTaylorMullen @rynowak 

I've also verified this change in design time in VS.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeParser.cs [hunk 1]
```diff
@@ -957,7 +957,27 @@ protected bool QualifiedIdentifier(out int identifierLength)
 
         protected bool NamespaceOrTypeName()
         {
-            if (Optional(CSharpSymbolType.Identifier) || Optional(CSharpSymbolType.Keyword))
+            if (Optional(CSharpSymbolType.LeftParenthesis))
+            {
+                while (!Optional(CSharpSymbolType.RightParenthesis) && !EndOfFile)
+                {
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeParser.cs]: `(   string val, int val2)` is a valid tuple. Therefore, need `Optional(CSharpSymbolType.WhiteSpace)` here (means you can 
 also remove the trailing one at the end of this while.

File: test/Microsoft.AspNetCore.Razor.Language.Test/Legacy/CSharpDirectivesTest.cs [hunk 2]
```diff
@@ -811,6 +811,30 @@ public void DirectiveDescriptor_AllowsNullableTypes(string expectedType)
                     Factory.Span(SpanKindInternal.Code, expectedType, markup: false).AsDirectiveToken(descriptor.Tokens[0])));
         }
 
+        [Theory]
+        [InlineData("(int aa, string bb)?")]
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Language.Test/Legacy/CSharpDirectivesTest.cs]: Add a spaced out tuple case: 
```
(    int     aa     ,    string     bb    )
```

**Comment by ajaybhargavb** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeParser.cs (hunk 1 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeParser.cs]: Added.

File: test/Microsoft.AspNetCore.Razor.Language.Test/Legacy/CSharpDirectivesTest.cs (hunk 2 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Razor.Language.Test/Legacy/CSharpDirectivesTest.cs]: Added.

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeParser.cs [hunk 3]
```diff
@@ -957,7 +957,28 @@ protected bool QualifiedIdentifier(out int identifierLength)
...
+                    {
+                        return false;
+                    }
+
+                    Optional(CSharpSymbolType.WhiteSpace);
+                    Optional(CSharpSymbolType.Identifier);
+                    Optional(CSharpSymbolType.WhiteSpace);
+                    Optional(CSharpSymbolType.Comma);
+                }
+
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeParser.cs]: Oops, missed this. Can have optional whitespace before the questionmark too. That's a little more tricky though. You only want to consume the whitespace if the next symbol is a question mark. Should do the same thing with our other `nullable` support below.

Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeParser.cs]: Huh?! Didn't realize that is valid.

Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeParser.cs]: Looks like even `int   ?aa = null;` is valid.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeParser.cs]: Yep, this is why languages don't usually treat whitespace as a token.

**Pull request merged**